City-of-Helsinki/hel-fi-drupal-grants
# feat: AU-2216, E2E: Inline form error validation

# [AU-2216](https://helsinkisolutionoffice.atlassian.net/browse/AU-2216)

## What was done

This pull request implements the possibility to validate inline form errors in application forms. Inline form error validation can be added to any form variant in a similar manner to normal error validation, by specifying a value for the `expectedInlineErrors` key:

```
const missingValues: FormDataWithRemoveOptionalProps = {
  title: 'Missing values',
  viewPageSkipValidation: true,
  formPages: {
    '1_hakijan_tiedot': {
      items: {},
      itemsToRemove: [
        'edit-bank-account-account-number-select',
        'edit-email',
        'edit-contact-person',
        'edit-contact-person-phone-number',
        'edit-community-address-community-address-select'
      ],
      expectedInlineErrors: [
        { selector: '.form-item-bank-account-account-number-select', errorMessage: 'Valitse tilinumero kenttä on pakollinen.' },
        { selector: '.form-item-email', errorMessage: 'Sähköpostiosoite kenttä on pakollinen.' },
        { selector: '.form-item-contact-person', errorMessage: 'Yhteyshenkilö kenttä on pakollinen.' },
        { selector: '.form-item-contact-person-phone-number', errorMessage: 'Puhelinnumero kenttä on pakollinen.' },
        { selector: '.webform-type-community-address-composite', errorMessage: 'Yhteisön osoite kenttä on pakollinen.' },
      ],
    },
...
```

The `expectedInlineErrors` key expects an array of `ExpectedInlineError` objects. Each object should specify a class selector for the `selector` key from which the error can be found on the form, and an expected error message inside the `errorMessage` key.

The process of validating the inline errors is relatively simple; While a form variant is being filled, we simply refresh the page by clicking the active progress step in the forms navigation. The refresh is performed AFTER all the desired form fields have been filled on the page, which triggers any inline error messages for form fields which have either empty or wrong values.

![Screenshot 2024-06-12 at 14 58 45](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/3492cea3-3794-493b-be7f-b2f3d1d73047)

A few examples of inline form error validation has been added to the `wrong_values` variant in `62`, and `missing_values` variant in `51`.


## How to install

Make sure your instance is up and running on correct branch.
* `feature/AU-2216-e2e-inline-form-error-validation`
* `make fresh`
* `make drush-cr`

## How to test

- Remove `ENABLED_FORM_VARIANTS` and `DISABLED_FORM_VARIANTS` from the `.env` file.

- Run `make test-pw-p PROJECT=forms-62 forms-51` and verify that you see the following output during the tests:

![Screenshot 2024-06-12 at 15 03 26](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/24d68ffa-07ed-4670-a32e-bc1f192ed80f)

![Screenshot 2024-06-12 at 15 03 16](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/8545874b-2de9-4368-aaaa-16b35fd88fd2)

![Screenshot 2024-06-12 at 15 03 03](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/f2fc49da-96b0-41fd-a46e-00e94ba4e13a)




[AU-2216]: https://helsinkisolutionoffice.atlassian.net/browse/AU-2216?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
